segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageManager.java

```diff
@@ -0,0 +1,1325 @@
+        this.config = Preconditions.checkNotNull(config, "config");
+        this.chunkStorage = Preconditions.checkNotNull(chunkStorage, "chunkStorage");
+        this.metadataStore = Preconditions.checkNotNull(metadataStore, "metadataStore");
+        this.executor = Preconditions.checkNotNull(executor, "executor");
+        this.closed = new AtomicBoolean(false);
+    }
+
+    /**
+     * Initializes the ChunkStorageManager.
+     *
+     * @param metadataStore Metadata store.
+     * @param containerId container id.
+     * @param systemJournal SystemJournal that keeps track of changes to system segments and helps with bootstrap.
+     * @throws Exception In case of any errors.
+     */
+    public void initialize(int containerId, ChunkMetadataStore metadataStore, SystemJournal systemJournal) throws Exception {
+        this.containerId = containerId;
+        this.metadataStore = Preconditions.checkNotNull(metadataStore, "metadataStore");
+        this.systemJournal = Preconditions.checkNotNull(systemJournal, "systemJournal");
+    }
+
+    /**
+     * Initializes this instance with the given ContainerEpoch.
+     *
+     * @param containerEpoch The Container Epoch to initialize with.
+     */
+    @Override
+    public void initialize(long containerEpoch) {
+        this.epoch = containerEpoch;
+    }
+
+
+    /**
+     * Attempts to open the given Segment in read-write mode and make it available for use for this instance of the Storage
+     * adapter.
+     * A single active read-write SegmentHandle can exist at any given time for a particular Segment, regardless of owner,
+     * while a read-write SegmentHandle can coexist with any number of read-only SegmentHandles for that Segment (obtained
+     * by calling openRead()).
+     * This can be accomplished in a number of different ways based on the actual implementation of the Storage
+     * interface, but it can be compared to acquiring an exclusive lock on the given segment).
+     *
+     * @param streamSegmentName Name of the StreamSegment to be opened.
+     * @return A CompletableFuture that, when completed, will contain a read-write SegmentHandle that can be used to access
+     * the segment for read and write activities (ex: read, get, write, seal, concat).
+     * If the segment is sealed, then a Read-Only handle is returned.
+     * <p>
+     * If the operation failed, it will be failed with the cause of the failure. Notable exceptions:
+     * <ul>
+     * <li> StreamSegmentNotExistsException: When the given Segment does not exist in Storage.
+     * </ul>
+     */
+    @Override
+    public CompletableFuture<SegmentHandle> openWrite(String streamSegmentName) {
+        checkInitialized();
+        return execute( () -> {
+            long traceId = LoggerHelpers.traceEnter(log, "openWrite", streamSegmentName);
+            Preconditions.checkNotNull(streamSegmentName, "streamSegmentName");
```

Comment:
Maybe the preconditions can be checked before event executing anything here.

Reply:
I think I had tried that very initially, but some of the callers can only handle CompletionException. Not worth making all that change for optimizing exceptional cases.